torchtext/csrc/gpt2_bpe_tokenizer.cpp

```diff
@@ -1,68 +1,318 @@
+    encoded.push_back(byte_encoder_.at((unsigned char)ch));
+  }
+  return encoded;
+}
+
+int64_t GPT2BPEEncoder::GetBPEMergeRank_(std::string pair) {
+  if (bpe_merge_ranks_.contains(pair)) {
+    return bpe_merge_ranks_.at(pair);
+  }
+  return inf_;
+}
+
+std::string GPT2BPEEncoder::FindBestPair_(std::vector<std::string> pairs) {
+  // Equivalent to:
+  //    min(pairs, key = lambda pair: self.bpe_merge_ranks.get(pair,
+  //    float('inf')))
+  auto best_pair_idx = 0;
+  auto best_rank = GetBPEMergeRank_(pairs[best_pair_idx]);
+
+  for (std::size_t i = 1; i < pairs.size(); ++i) {
+    auto rank = GetBPEMergeRank_(pairs[i]);
+    if (rank < best_rank) {
+      best_pair_idx = i;
+      best_rank = rank;
+    }
+  }
+  return pairs[best_pair_idx];
+}
+
+std::vector<std::string> GPT2BPEEncoder::BPE_(
+    const std::vector<std::string> &token_list) {
+  // Given a list of input tokens, keep finding the best bpe merge and
+  // generate a new list of tokens until
+  //  1) token list size reduced to 1
+  //      OR
+  //  2) can't find bpe merge
+  auto concatenated = _concatenate_strings(token_list);
+  if (caching_enabled_ && cache_.contains(concatenated)) {
+    return cache_.at(concatenated);
+  }
+
+  std::vector<std::string> tok_list = token_list;
+  auto pairs = _get_pairs(tok_list, seperator_);
+  if (pairs.empty()) {
+    return tok_list;
+  }
+  while (true) {
+    auto bigram = FindBestPair_(pairs);
+    if (!bpe_merge_ranks_.contains(bigram)) break;
+
+    // Finding all indexes that token_list[i] == first and token_list[i+1] ==
+    // second. After the loop, new token list will be
+    //  1) first + second pair
+    //  2) all the other tokens in the original token list
+    //
+    // For example: first="a" second="w" and token_list =
+    // ["a", "w", "some", "a", "w", "e"]
+    // Result: new_token_list = ["aw", "some", "aw", "e"]
+
+    auto parts = _split_tokens(bigram, seperator_);
+    std::vector<std::string> new_token_list;
+    std::size_t i = 0;
+    while (i < tok_list.size()) {
+      auto j = _list_str_index(tok_list, parts.first, i);
+      if (j != -1) {
+        for (int k = i; k < j; k++) new_token_list.push_back(tok_list[k]);
+        i = j;
+      } else {
+        for (std::size_t k = i; k < tok_list.size(); k++)
+          new_token_list.push_back(tok_list[k]);
+        break;
+      }
+
+      if (tok_list[i] == parts.first && i < (tok_list.size() - 1) &&
+          tok_list[i + 1] == parts.second) {
+        new_token_list.push_back(parts.first + parts.second);
+        i += 2;
+      } else {
+        new_token_list.push_back(tok_list[i]);
+        i += 1;
+      }
+    }
+
+    tok_list = new_token_list;
+    if (tok_list.size() == 1) {
+      break;
+    } else {
+      pairs = _get_pairs(tok_list, seperator_);
+    }
+  }
+
+  if (caching_enabled_) cache_.insert(concatenated, tok_list);
+  return tok_list;
+}
+
+std::vector<int64_t> GPT2BPEEncoder::Encode(const std::string &text) {
```

Comment:
We could instead use C10::string_view to avoid copying the strings. [Here](https://github.com/pytorch/text/blob/be7bc66e9d720850850ce8551af021c957cadf03/torchtext/csrc/vocab.cpp#L35) is how it is done for vocab. Then we can register like [here](https://github.com/pytorch/text/blob/be7bc66e9d720850850ce8551af021c957cadf03/torchtext/csrc/register_pybindings.cpp#L119) for pybind to potentially gain performance.

Reply:
That sounds good. I can take this in a follow up PR as there will be few code changes in some private methods as well, including the interface of the `gpt2_bpe_pre_tokenizer ` method. Adding a follow up item for the same!